server/api/cvrs.py

```diff
@@ -0,0 +1,241 @@
+                        contests_metadata[contest_name]["choices"][contest_choice][
+                            "num_votes"
+                        ] += int(interpretation)
+                        contests_on_ballot.add(contest_name)
+                for contest_name in contests_on_ballot:
+                    contests_metadata[contest_name]["total_ballots_cast"] += 1
+
+            jurisdiction.cvr_contests_metadata = contests_metadata
+
+            # In order to use COPY, we have to bypass SQLAlchemy and use
+            # the underlying DBAPI (psycogp2). This means these commands
+            # will happen in a separate transaction from the surrounding
+            # context.
+            connection = db_engine.raw_connection()
+            try:
+                cursor = connection.cursor()
+                cursor.execute("BEGIN")
+                cursor.execute(
+                    f"""
+                        DELETE FROM cvr_ballot
+                        WHERE batch_id IN (
+                            SELECT id FROM batch
+                            WHERE jurisdiction_id = '{jurisdiction.id}'
+                        )
+                        """
+                )
+                ballots_tempfile.seek(0)
+                cursor.copy_expert(
+                    """
+                        COPY cvr_ballot
+                        FROM STDIN
+                        WITH (
+                            FORMAT CSV,
+                            DELIMITER ',',
+                            HEADER
+                        )
+                        """,
+                    ballots_tempfile,
+                )
+                cursor.execute("COMMIT")
+                cursor.close()
+                connection.commit()
+            except Exception as exc:
+                cursor.execute("ROLLBACK")
+                raise exc
+            finally:
+                connection.close()
+
+    process_file(session, file, process)
+
+
+# Raises if invalid
+def validate_cvr_upload(request: Request):  # pragma: no cover
+    # TODO test
+    if "cvr" not in request.files:
+        raise BadRequest("Missing required file parameter 'cvr'")
+
+
+# We save the CVR file, and bgcompute finds it and processes it in
+# the background.
+def save_cvr_file(cvr, jurisdiction: Jurisdiction):
+    cvr_string = decode_csv_file(cvr.read())
+    jurisdiction.cvr_file = File(
+        id=str(uuid.uuid4()),
+        name=cvr.filename,
+        contents=cvr_string,
+        uploaded_at=datetime.utcnow(),
+    )
+
+
+def clear_cvr_file(jurisdiction: Jurisdiction, delete_cvrs: bool = True):
+    if jurisdiction.cvr_file_id:
+        File.query.filter_by(id=jurisdiction.cvr_file_id).delete()
+    if delete_cvrs:  # pragma: no cover
+        # TODO test
+        CvrBallot.query.filter_by(jurisdiction_id=jurisdiction.id).delete()
+
+
+@api.route(
+    "/election/<election_id>/jurisdiction/<jurisdiction_id>/cvr", methods=["PUT"],
```

Comment:
probably should be `cvrs` with an `s`, because CVR really means a single vote. But I can absolutely live with this pedantic comment being ignored :)

Reply:
no i love this. i realized after using the acronym a bunch of places it was not clear to me whether "record" referred to a whole file or a single row